src/state/content_seed_client.rs

```diff
@@ -0,0 +1,72 @@
+use candid::Principal;
+use reqwest::Client;
+use reqwest::Url;
+use serde::Deserialize;
+use serde::Serialize;
+use std::error::Error;
+
+use crate::auth::DelegatedIdentityWire;
+
+#[derive(Deserialize)]
+pub struct AllowPrincpalRes {
+    allowed: bool,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct UploadContentPayload {
+    url: String,
+    payload: DelegatedIdentityWire,
+}
+
+#[derive(Deserialize)]
+pub struct ServerError {
+    error: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct ContentSeedClient {
+    client: Client,
+    base_url: Url,
+}
+
+impl ContentSeedClient {
+    pub fn with_base_url(url: Url) -> Self {
+        ContentSeedClient {
+            client: Default::default(),
+            base_url: url,
+        }
+    }
+
+    pub async fn check_if_authorized(&self, principal: Principal) -> Result<bool, Box<dyn Error>> {
+        let api_url = self
+            .base_url
+            .join("allowed/")
+            .expect("url error")
+            .join(&principal.to_string())
+            .expect("url error");
+        let res = self.client.get(api_url).send().await?;
+        let res_json: AllowPrincpalRes = res.json().await?;
+        Ok(res_json.allowed)
+    }
+
+    pub async fn upload_content(
+        &self,
+        url: String,
+        identity: DelegatedIdentityWire,
+    ) -> Result<(), Box<dyn Error>> {
```

Comment:
prefer `ServerFnError` instead of `Box<dyn Error>`
More ideally, i'd like fully typed errors

Reply:
What additional benefit do we have mentioning ServerFnError we cannot recover from it and we will just print out the message from the error so doesn't make sense to use strongly typed error.

Reply:
The added benefit is consistency, the leptos ecosystem uses ServerFnError for arbitrary errors and so does the code base.
Why does function XYZ use ServerFnError but ABC use Box<dyn Error>?

Strong typing is always preferred, in the future we might have better error monitoring and proper types give free information about what the error is.

Reply:
Consistency is not an issue here we can move every api to return Box<dyn Error>. Strong typing is preferred if you can actually recover form the error or have some additional logic for error type. Monitoring would log the error as far as I know and using typed erased error when you just want to know the cause is preferred.

Reply:
Consistency is an issue for the code base, I dont see why we'd want ti deviate from what we already do.
Its similar to how a crate that uses anyhow wouldnt use Box<dyn Error>.

Secondly,
You don't know the future and neither do I, We dont know what the monitoring API will look liike, we dont know how we will be handling errors in the future. Strong typing is a defensive mechanism that will cause us less headaches in the future. If you have really strong opinions against strongly typed errors, i dont really mind, we can add the types whenever required, but it has been discussed well before and multiple times that we should prefer strong types in the leptos code base (cc: @s-go-bazzinga @komal-rs )

Reply:
I don't mind. I just wrote what was suggested by Jon Gjengset in one his book.

_In general, the community consensus is that errors should be rare and therefore should not add much cost to the "happy path." For that reason, errors are often placed behind a pointer type, such as a Box or Arc. This way, they're unlikely to add much to the size of the overall Result type they're contained within._

Reply:
Overall, always prefer strong typing wherever possible. You get meagre storage benefits in terms of lesser sized compile outputs but drastically reduced safety/compile time guarantees. And more CPU usage as well due to dynamic inferencing.

Storage is cheap, compute almost always costlier.

Finally make the compiler do as much as possible. The lesser a dev needs to remember to "jump", the lesser chances they're going to trip..

My 2 cents.

Reply:
Agreed. 
Based on some searches, looks like this is an active debate in the Rust community with even some performant crates like reqwest and ripgrep going the other route of dyn - https://users.rust-lang.org/t/result-type-when-to-box-vs-when-to-wrap/59084 . 

I think the important thing is for us to agree on one approach and be consistent with it as much as possible, atleast localized to relevant repos.

Reply:
I think type erased doesn't mean there would be no compile time checks. The only idea is errors are rare cases and if you don't want to handling based on error types and just want to know the cause it is okay to just send `dyn Error` which provides the reason but doesn't provide data that is not needed.

Reply:
Chiming in again. Those crates are library crates that will have hundreds of eyeballs validating that there are no errors and millions of dependent crates that are battle testing in production that everything runs as expected. In our case, we don't have that privilege and we wan't to get as many compiler guarantees regarding safety as possible.

Also Boxed generics are ugly as heck for the reader

Reply:
I am just confused because I don't find a geniuene case over here include types. Box generics in erros help in developer experience by returning any type of error that implements error trait that is the only useful information that is needed as even in error case we are printing the error cause. What benefit does Typed Error type enum/struct provide in these use case over Boxed type. I have two types of error in the same block one retrieved from json and one from request itself. If I use typed I need to return an enum that implements Error but given that I will not be using that anywhere in the code what is the use? Plus `Box<dyn Error>` also gurantees you will only be able to access methods in trait Error which is all that will be needed for logging.